File: 백준/Gold/1865. 웜홀/웜홀.py
import sys
from collections import Counter

tc = int(sys.stdin.readline())

for i in range(tc):
    node, vertex, warm = map(int, sys.stdin.readline().split())
    lis = []
    for j in range(vertex):
        temp = list(map(int, sys.stdin.readline().split()))
        lis.append((temp[0], temp[1], temp[2]))
        lis.append((temp[1], temp[0], temp[2]))
    for k in range(warm):
        temp = list(map(int, sys.stdin.readline().split()))
        lis.append((temp[0], temp[1],-temp[2]))

    # 시작지점
    key = False

    dist = [10001 for q in range(node+1)]
    dist[1] = 0
    for e in range(node):
        for v in range(vertex * 2 + warm):
            now = lis[v][0]
            next = lis[v][1]
            weight = lis[v][2]
            if dist[next] > dist[now] + weight:
                dist[next] = dist[now] + weight
                if e == node - 1:
                    key = True

    if not key:
        print("NO")
    else:
        print("YES")

File: 백준/Gold/1865. 웜홀/웜홀.java
import java.io.*;
import java.util.*;

public class Main {

    static int fois;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        fois = Integer.parseInt(br.readLine());
        for(int i = 0; i < fois; i ++){
            ArrayList<List<Integer>> lis = new ArrayList<>();
            int n,m,w;
            int[] tmp = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            n = tmp[0];
            m = tmp[1];
            w = tmp[2];
            for(int j = 0; j < m; j++){
                int[] tmp1 = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
                lis.add(List.of(tmp1[0],tmp1[1],tmp1[2]));
                lis.add(List.of(tmp1[1],tmp1[0],tmp1[2]));
            }
            for(int j = 0; j < w; j++){
                int[] tmp1 = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
                lis.add(List.of(tmp1[0],tmp1[1],-tmp1[2]));
            }
            int[] dist = new int[n+1];
            for(int j = 1; j <= n; j ++){
                dist[j] = 10001;
            }
            dist[1] = 0;
            boolean key = true;
            for(int j = 1; j <= n; j++){
                for(int l = 0; l < lis.size(); l++){
                    int now = lis.get(l).get(0);
                    int nex = lis.get(l).get(1);
                    int wei = lis.get(l).get(2);
                    if(dist[nex] > dist[now] + wei){
                        dist[nex] = dist[now] + wei;
                        if(j == n){
                            key = false;
                        }
                    }
                }
            }
            if(key){
                bw.write("NO" + "\n");
            }
            else{
                bw.write("YES" + "\n");
            }
        }

        bw.flush();
        bw.close();
        br.close();
    }
}